File: code-awesome/go-gin-react/webshop2/backend/src/components/product-list.go
package main

import (
	"fmt"
	"log"
	"net/http"

	"github.com/gin-gonic/gin"
)

type Product struct {
	ID        int
	Title     string
	Image     string
	Price     float64
	DescPart  string
	Category string
}

// Mock database (replace with actual database interaction)
var products = []Product{
	{ID: 1, Title: "Cute Kitten Toy", Image: "/images/kitten.jpg", Price: 19.99, DescPart: "Soft and cuddly...", Category: "Toys"},
	{ID: 2, Title: "Fluffy Bunny Plush", Image: "/images/bunny.jpg", Price: 24.99, DescPart: "A super soft bunny...", Category: "Toys"},
	{ID: 3, Title: "Rainbow Ball", Image: "/images/ball.jpg", Price: 9.99, DescPart: "Perfect for fetch!", Category: "Toys"},
	{ID: 4, Title: "Cozy Cat Bed", Image: "/images/bed.jpg", Price: 39.99, DescPart: "A warm and comfy bed...", Category: "Furniture"},
	{ID: 5, Title: "Playful Puppy Puzzle", Image: "/images/puppy.jpg", Price: 29.99, DescPart: "Keeps your puppy entertained...", Category: "Toys"},
}

// productListHandler handles the request to list products.
func productListHandler(c *gin.Context) {
	// Simulate pagination (for demonstration purposes)
	page := 1
	limit := 10

	// Filter products by category (simulated)
	var filteredProducts []Product
	category := c.DefaultQuery("category", "Toys") // Get category from query parameters
	for _, product := range products {
		if product.Category == category {
			filteredProducts = append(filteredProducts, product)
		}
	}

	// Simulate pagination (skip and take)
	startIndex := (page - 1) * limit
	endIndex := startIndex + limit
	if endIndex > len(filteredProducts) {
		endIndex = len(filteredProducts)
	}

	// Return the paginated list of products
	c.JSON(200, filteredProducts[startIndex:endIndex])
}

func main() {
	router := gin.Default()
	router.GET("/products", productListHandler)
	router.Run(":8080")
}

File: code-awesome/java-quarkus-vue/webshop2/frontend/src/components/product-list.js
// product-list.js
import { reactive } from 'vue';

// Simulated data for products (replace with actual database interaction)
const allProducts = [
  { id: 1, title: 'Cute Puppy Toy', image: 'puppy.jpg', price: 19.99, description: 'A squeaky toy for your furry friend.' },
  { id: 2, title: 'Fluffy Kitten Plush', image: 'kitten.jpg', price: 24.99, description: 'A soft and cuddly kitten plush.' },
  { id: 3, title: 'Squirrel Figurine', image: 'squirrel.jpg', price: 12.50, description: 'A charming squirrel figurine.' },
  { id: 4, title: 'Bunny Stuffed Animal', image: 'bunny.jpg', price: 15.75, description: 'A cuddly bunny plush toy.' },
];

// Reactive data for the product list
const productList = reactive({
  products: allProducts,
  currentCategory: 'all', // Default category
  pageSize: 5,
  currentPage: 1,
});

// Function to filter products by category
const filterProducts = (category) => {
  productList.currentCategory = category;
  if (category === 'all') {
    productList.products = allProducts;
  } else {
    productList.products = allProducts.filter(product => product.category === category);
  }
};

// Function to handle pagination
const goToPage = (pageNumber) => {
  productList.currentPage = pageNumber;
  productList.products = []; // Reset products for the new page
  for (let i = (pageNumber - 1) * pageSize; i < Math.min(pageNumber * pageSize, allProducts.length); i++) {
    productList.products.push(allProducts[i]);
  }
};

// Expose reactive data and functions to the component
export default {
  data() {
    return {
      productList,
    };
  },
  methods: {
    filterCategory: filterCategory,
    goToPage: goToPage,
  },
};